src/nvim/tui/tui.c

```diff
@@ -463,33 +539,74 @@ static void tui_mouse_off(UI *ui)
   data->mouse_enabled = false;
 }
 
+static void tui_set_cursor(cursorentry_T c)
+{
+  int shape = c.shape;
+  bool inside_tmux = os_getenv("TMUX") != NULL;
+
+# define TMUX_WRAP(seq) (inside_tmux ? "\x1bPtmux;\x1b" seq "\x1b\\" : seq)
+  // Support changing cursor shape on some popular terminals.
+  const char *term_prog = os_getenv("TERM_PROGRAM");
+  const char *vte_version = os_getenv("VTE_VERSION");
+
+  if ((term_prog && !strcmp(term_prog, "Konsole"))
+      || os_getenv("KONSOLE_DBUS_SESSION") != NULL) {
+    // Konsole uses a proprietary escape code to set the cursor shape
+    // and does not support DECSCUSR.
+    switch (shape) {
+      case SHAPE_VER: shape = 1; break;
+      case SHAPE_HOR: shape = 3; break;
+      case SHAPE_BLOCK: shape = 0; break;
+      default: WLOG("Unknown shape value %d", shape); break;
+    }
+    printf(TMUX_WRAP("\x1b]50;CursorShape=%d;BlinkingCursorEnabled=%d\x07"),
+           shape, (c.blinkon !=0));
+  } else if (!vte_version || atoi(vte_version) >= 3900) {
+    // Assume that the terminal supports DECSCUSR unless it is an
+    // old VTE based terminal.  This should not get wrapped for tmux,
+    // which will handle it via its Ss/Se terminfo extension - usually
+    // according to its terminal-overrides.
+
+    switch (shape) {
+      case SHAPE_VER: shape = 5; break;
+      case SHAPE_HOR: shape = 3; break;
+      case SHAPE_BLOCK: shape = 1; break;
+      default: WLOG("Unknown shape value %d", shape); break;
+    }
+
+    // odd => blink ?
+    ILOG("x1b[%d q (blinking= %d)", shape + (c.blinkon ==0), (c.blinkon != 0));
+    printf("\x1b[%d q", shape + (c.blinkon ==0));
```

Comment:
I removed the ILOG statement but I think the printf is in the   `else if (!vte_version || atoi(vte_version) >= 3900) {` block. Maybe I shouldn't use printf but some unib_* call instead ?